src/tile-plotting.jl

```diff
@@ -0,0 +1,281 @@
+        sort!(available_to_remove, by=tile-> abs(tile.z - m.zoom[]))
+        n_avail = length(available_to_remove)
+        to_remove = available_to_remove[1:(n_avail - 200)]
+        for tile in to_remove
+            plot_key = remove_unused!(m, tile)
+            if !isnothing(plot_key)
+                remove_plot!(m, plot_key[2])
+            end
+        end
+    end
+
+    if isempty(m.unused_plots)
+        mplot = create_tileplot!(cfg, m.axis, data_processed, bounds, (tile, m.crs))
+    else
+        mplot = pop!(m.unused_plots)
+        update_tile_plot!(mplot, cfg, m.axis, data_processed, bounds, (tile, m.crs))
+    end
+
+    if haskey(m.current_tiles, tile)
+        move_in_front!(mplot, abs(m.zoom[] - tile.z), bounds)
+    else
+        move_to_back!(mplot, abs(m.zoom[] - tile.z), bounds)
+    end
+
+    # Always move new plots to the front
+    mplot.visible = true
+    cfg.postprocess(mplot)
+    m.plots[key] = (mplot, tile, bounds)
+    return
+end
+
+############################
+#### Elevation Data plotting
+####
+
+struct ElevationData
+    elevation::AbstractMatrix{<: Number}
+    color::AbstractMatrix{<: Colorant}
+end
+
+function Base.map(f::Function, data::ElevationData)
+    return ElevationData(
+        map(f, data.elevation),
+        data.color
+    )
+end
+
+function get_bounds(tile::Tile, data::ElevationData, crs)
+    ext = MapTiles.extent(tile, crs)
+    mini, maxi = extrema(data.elevation)
+    origin = Vec3d(ext.X[1], ext.Y[1], mini)
+    w = Vec3d(ext.X[2] - ext.X[1], ext.Y[2] - ext.Y[1], maxi - mini)
+    return Rect3d(origin, w)
+end
+
+function create_tileplot!(config::PlotConfig, axis::AbstractAxis, data::ElevationData, bounds::Rect, tile_crs)
+    # not so elegant with empty array, we may want to make this a bit nicer going forward
+    color = isempty(data.color) ? (;) : (color=data.color,)
+    mini, maxi = extrema(bounds)
+    p = Makie.surface!(
+        axis.scene,
+        (mini[1], maxi[1]), (mini[2], maxi[2]), data.elevation;
+        color...,
+        shading=Makie.NoShading,
+        inspectable=false,
+        config.attributes...
+    )
+    return p
+end
+
+function update_tile_plot!(plot::Surface, ::PlotConfig, ::AbstractAxis, data::ElevationData, bounds::Rect, tile_crs)
+    mini, maxi = extrema(bounds)
+    plot.args[1].val = (mini[1], maxi[1])
+    plot.args[2].val = (mini[2], maxi[2])
+    plot[3] = data.elevation
+    if !isempty(data.color)
+        plot.color = data.color
+    end
+    return
+end
+
+############################
+#### Image Data plotting
+####
+
+const ImageData = AbstractMatrix{<:Colorant}
+
+function create_tileplot!(config::PlotConfig, axis::AbstractAxis, data::ImageData, bounds::Rect, tile_crs)
+    mini, maxi = extrema(bounds)
+    plot = Makie.image!(
+        axis.scene,
+        (mini[1], maxi[1]), (mini[2], maxi[2]), rotr90(data);
+        inspectable=false,
+        config.attributes...
+    )
```

Comment:
I guess this will have to be overridden with a sampled mesh for e.g. GeoAxis / PolarAxis?

Reply:
Yeah, that's what you can do now :)